fix(nix/devShells): use mold as rust's default linker and ensure functional openssl linking

steveej opened the pull request:
### Summary

fixes #2937

* changed how the holonix integration tests work and made it more realistic to reflect end-user environments.
* i added a minimum example crate to verify that openssl can be linked statically with 


### TODO:
- [ ] CHANGELOG(s) updated with appropriate info
- [ ] Just before pressing the merge button, ensure new entries to CHANGELOG(s) are still under the _UNRELEASED_ heading

steveej commented:
this breaks `nix` on my system within the nix shell:

```
nix: /nix/store/whypqfa83z4bsn43n4byvmw80n4mg3r8-glibc-2.37-45/lib/libc.so.6: version `GLIBC_2.38' not found (required by /nix/store/vzajrlhsdv2d39s7v6zv09ggajs05gwj-openssl-3.0.11/lib/libcrypto.so.3)
nix: /nix/store/whypqfa83z4bsn43n4byvmw80n4mg3r8-glibc-2.37-45/lib/libc.so.6: version `GLIBC_2.38' not found (required by /nix/store/vzajrlhsdv2d39s7v6zv09ggajs05gwj-openssl-3.0.11/lib/libcrypto.so.3)
```

it probably breaks everything that links to any library of the `openssl` package.

steveej commented on nix/modules/devShells.nix:
```diff
@@ -172,26 +170,34 @@
-            };
+          packages = (lib.lists.optionals pkgs.stdenv.isLinux [
+            pkgs.mold
+            pkgs.pkgsStatic.openssl
```
i'm not clear on why this is necessary to add explicitly. it should be introduced by having the holochain package in `inputsFrom`. 

while i'm not clear on why exactly it works i can confirm it does and so i accept it as a mysterious yet working solution.

ThetaSinner replied:
This is more serious for people who are building on Windows. It's multiple error-prone steps to get OpenSSL provided and found for linking on Windows. If something has broken with our feature flags to ensure OpenSSL gets built as part of the Rust build then we need to follow up on that. Could you open a GitHub issue or similar please? I'm happy for it to be assigned to me as long as we are tracking it

jost-s reviewed (approved):
LGTM! Great to return to the stock mkShell.

I see that `mold` is no longer broken on darwin [since 22-11](https://github.com/NixOS/nixpkgs/blob/nixos-22.11/pkgs/development/tools/mold/default.nix#L39-L51). I'll try to set that up on my machine and make a PR then.

ThetaSinner commented on nix/modules/devShells.nix:
```diff
@@ -172,26 +170,34 @@
+            export DYLD_FALLBACK_LIBRARY_PATH="$(rustc --print sysroot)/lib"
+          '')
+          + (lib.strings.optionalString pkgs.stdenv.isLinux ''
+            export RUSTFLAGS="$RUSTFLAGS -Clink-arg=-fuse-ld=mold"
```
Does this inherit Rust flags from outside the Nix environment or is this just able to inherit flags set elsewhere by us?

steveej replied:
> Does this inherit Rust flags from outside the Nix environment

yes:

```
$ env RUSTFLAGS="outside" nix develop .#coreDev
Rust development shell spawned. Type exit to leave.
pre-commit-hooks.nix: hooks up to date

[coreDev:~/src/holo/holochain]$ echo $RUSTFLAGS
outside -Clink-arg=-fuse-ld=mold
```

ThetaSinner replied:
Okay cool, that's something to be aware of then. Thanks for demonstrating :)

steveej replied:
my rationale is that the user can prevent any environment variables from being passed into the `nix develop` shell:

```
      · --ignore-environment / -i

        Clear the entire environment (except those specified with --keep).
```

so keeping the possibility of actually using this one instead of clobbering it.

ThetaSinner commented:
Is all this automatically inherited when `mkShell` is used in Holonix? Because that's the problem I had with the fix I did. That settings in the parent shell aren't all inherited so you end up with Holonix working correctly from the Holochain repo but not once you import it and create a new shell from it

ThetaSinner commented:
and related, is there anything we can do to automatically test this? That's probably for a follow-up piece of work though :)

steveej commented:
@ThetaSinner 

> Is all this automatically inherited when `mkShell` is used in Holonix?

yes, if [`inputsFrom` is used](https://github.com/NixOS/nixpkgs/blob/c2892b3fc93bd5cce8f72d2aa97287a870fa05d2/pkgs/build-support/mkshell/default.nix#L36-L42). this PR does make use of that since the changes go into `rustDev` which `holonix` inherits from via `inputsFrom`.


> is there anything we can do to automatically test this?

yes, we could add a test for this. i didn't add one that compiles something within holonix but i do have a minimum reproduction for this. i could add it and see how much time it adds to the integration check. follow-up or here?

steveej commented:
> i could add it and see how much time it adds to the integration check. follow-up or here?

i added a test here, and i changed how the test uses holonix from overriding it to creating a new shell that refers to it in `inputsFrom`. this is closer to how we instruct users.

ThetaSinner reviewed (approved):
Nice, looks great :)
